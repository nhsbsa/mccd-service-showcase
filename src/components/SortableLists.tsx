'use client';

import React, { useCallback, useEffect, useState } from 'react';
import * as _ from 'lodash';
import { GridRow, GridColumn } from '@/components/GridLayout';
import { DataItem, DataType } from '@/types';
import Link from 'next/link';
import { kebabCase } from 'lodash';

type TSortableList = {
  sortBy?: DataItem
  titleField?: string
  linkField?: string
  sortText?: string
  data?: DataType
}
function SortableLists({
  sortBy = {},
  sortText = 'Sort by',
  data = [],
  titleField = 'title',
  linkField = '',
}:TSortableList) {
  const [listData, setListData] = useState<DataType>([]);
  // eslint-disable-next-line max-len
  const [searchResults, setSearchResults] = useState<Array<[string, { slug: string[]; title: string }[]]>>([]);

  const groupBy = useCallback(<K extends keyof DataItem>(d: DataItem[], g: K) => {
    if (!(d?.length > 0)) return;
    const groupedData = d.reduce((accumulator, item) => {
      const category = item[g] as string;
      if (category !== undefined && typeof category === 'string') {
        const cat: string = category.toLowerCase();
        const group: any = accumulator[cat] || [];
        accumulator[cat] = [...group, item];
      }
      return accumulator;
    }, {} as Record<string, DataItem[]>);

    const sortOrder = (sortBy[g] as string[]) || [];
    const sorted = Object.entries(groupedData).sort(
      (a, b) => sortOrder.indexOf(a[0]) - sortOrder.indexOf(b[0]),
    );
    setSearchResults(sorted as Array<[string, { slug: string[]; title: string }[]]>);
  }, [sortBy]);

  useEffect(() => {
    if (data?.length > 0) setListData(data);
  }, [data]);

  useEffect(() => {
    const sortKey = Object.keys(sortBy)[0];
    groupBy(listData, sortKey);
  }, [listData, sortBy, groupBy]);

  const handleSelect = (e: React.ChangeEvent<HTMLSelectElement>) => {
    e.preventDefault();
    const sortValue = e.target.value;
    groupBy(listData, sortValue);
  };
  const makeSort = () => (
    <div key="sort" className="govuk-form-group">
      <label className="govuk-label" htmlFor="sort">
        {sortText}
      </label>
      <select className="govuk-select" id="sort" name="sort" onChange={handleSelect}>
        <option value="" disabled hidden>{sortText}</option>
        {Object.keys(sortBy).map((v) => {
          if (v.includes(':')) {
            const c = v.split(':');
            const value = c[1];
            const key = c[0];
            return <option key={v} value={value}>{_.capitalize(key)}</option>;
          }
          return <option key={v} value={_.camelCase(v)}>{_.capitalize(v)}</option>;
        })}
      </select>
    </div>
  );
  const makeLists = () => (searchResults ? (
    <GridRow>
      {searchResults.map((group, i) => (
        <GridColumn key={`group-${i + 1}`} width="1/4">
          <h3 className="govuk-heading-m">{_.capitalize(group[0] || '')}</h3>
          <ul className="govuk-list govuk-list--spaced govuk-!-padding-top-4">
            {group[1]
              .map((groupItem, index:number) => (
                <li key={`list-item-${index + 1}`}>
                  {groupItem[linkField]
                    ? <Link data-testid={kebabCase(groupItem[titleField])} href={`${groupItem[linkField]}`} title={groupItem[titleField]}>{groupItem[titleField]}</Link>
                    : groupItem[titleField]}
                </li>
              ))}
          </ul>
        </GridColumn>
      ))}
    </GridRow>
  )
    : null);
  return (listData?.length > 0
    ? (
      <section>
        {makeSort()}
        {makeLists()}
      </section>
    )
    : <p className="govuk-body-s">No data</p>
  );
}
export default SortableLists;
