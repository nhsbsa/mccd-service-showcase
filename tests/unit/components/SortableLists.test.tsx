import React from 'react';
import userEvent from '@testing-library/user-event';
import {
  screen, render, fireEvent,
} from '@testing-library/react';
import SortableLists from '@/components/SortableLists';

const sampleData = [
  { title: 'Item 1', category: 'A', link: '/Link1' },
  { title: 'Item 2', category: 'B', link: '/Link2' },
  { title: 'Item 3', category: 'A', link: '/Link3' },
  { title: 'Item 4', category: 'B', link: '/Link4' },
];

const sampleSortBy = {
  category: ['B', 'A'],
};

describe('SortableLists component', () => {
  it('renders the component without props and without crashing', () => {
    const { container } = render(
      <SortableLists />,
    );
    expect(container).toBeInTheDocument();
  });

  it('renders "No data" message when no data is provided', () => {
    render(<SortableLists sortBy={sampleSortBy} data={[]} linkField="link" />);
    expect(screen.getByText('No data')).toBeInTheDocument();
  });

  it('does not render "Sort by" label when no data is provided', () => {
    render(<SortableLists sortBy={sampleSortBy} data={[]} linkField="link" />);
    expect(screen.queryByLabelText('Sort by')).not.toBeInTheDocument();
  });

  it('renders grouped lists correctly', () => {
    render(<SortableLists sortBy={sampleSortBy} data={sampleData} linkField="link" />);
    expect(screen.getByText('A')).toBeInTheDocument();
    expect(screen.getByText('B')).toBeInTheDocument();
  });

  it('handles select change and regroups data', () => {
    render(<SortableLists sortBy={sampleSortBy} data={sampleData} linkField="link" />);
    const select = screen.getByLabelText('Sort by');
    fireEvent.change(select, { target: { value: 'category' } });
    expect(screen.getByText('B')).toBeInTheDocument();
    expect(screen.getByText('A')).toBeInTheDocument();
  });

  it('assigns correct links in makeLists', () => {
    render(<SortableLists sortBy={sampleSortBy} data={sampleData} linkField="link" />);

    sampleData.forEach(({ link, title }) => {
      const linkElement = screen.getByRole('link', { name: title });
      expect(linkElement).toBeInTheDocument();
      expect(linkElement).toHaveAttribute('href', link);
    });
  });

  it('handles select change and regroups data with correct links', async () => {
    render(<SortableLists sortBy={sampleSortBy} data={sampleData} linkField="link" />);
    const select = screen.getByLabelText('Sort by');
    await userEvent.selectOptions(select, 'category');

    sampleData.forEach(({ link, title }) => {
      const linkElement = screen.getByRole('link', { name: title });
      expect(linkElement).toBeInTheDocument();
      expect(linkElement).toHaveAttribute('href', link);
    });
  });

  // it('matches the snapshot', () => {
  //   const { asFragment } = render(
  //     <SortableLists sortBy={sampleSortBy} data={sampleData} />,
  //   );
  //   expect(asFragment()).toMatchSnapshot();
  // });
});
