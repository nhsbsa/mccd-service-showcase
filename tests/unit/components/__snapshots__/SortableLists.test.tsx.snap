// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortableLists component matches the snapshot 1`] = `
<DocumentFragment>
  <section>
    <div
      class="govuk-form-group"
    >
      <label
        class="govuk-label"
        for="sort"
      >
        Sort by
      </label>
      <select
        class="govuk-select"
        id="sort"
        name="sort"
      >
        <option
          disabled=""
          hidden=""
          value=""
        >
          Sort by
        </option>
        <option
          value="category"
        >
          Category
        </option>
      </select>
    </div>
    <div
      class="govuk-grid-row"
      data-testid="grid-row"
    >
      <div
        class="govuk-grid-column-one-quarter"
        data-testid="grid-column"
      >
        <h3
          class="govuk-heading-m"
        >
          A
        </h3>
        <ul
          class="govuk-list govuk-list--spaced govuk-!-padding-top-4"
        >
          <li>
            Item 1
          </li>
          <li>
            Item 3
          </li>
        </ul>
      </div>
      <div
        class="govuk-grid-column-one-quarter"
        data-testid="grid-column"
      >
        <h3
          class="govuk-heading-m"
        >
          B
        </h3>
        <ul
          class="govuk-list govuk-list--spaced govuk-!-padding-top-4"
        >
          <li>
            Item 2
          </li>
          <li>
            Item 4
          </li>
        </ul>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
